src/main/java/com/google/api/gax/grpc/CallContext.java

```diff
@@ -43,22 +44,36 @@
  * return copies of the object, but with one field changed.
  * The immutability and thread safety of the arguments solely depends on the arguments themselves.
  */
-public final class CallContext<RequestT> {
+public final class CallContext<RequestT, ResponseT> {
   private final Channel channel;
   private final CallOptions callOptions;
   private final RequestT request;
 
-  private CallContext(Channel channel, CallOptions callOptions, RequestT request) {
+  private final StreamObserver<ResponseT> responseObserver;
+
+  private CallContext(
+      Channel channel,
+      CallOptions callOptions,
+      RequestT request,
+      StreamObserver<ResponseT> responseObserver) {
     this.channel = channel;
     this.callOptions = callOptions;
     this.request = request;
+    this.responseObserver = responseObserver;
+  }
+
+  public static <RequestT, ResponseT> CallContext<RequestT, ResponseT> of(
+      Channel channel, CallOptions callOptions, RequestT request) {
+    return new CallContext<RequestT, ResponseT>(channel, callOptions, request, null);
   }
 
-  public static <T> CallContext<T> of(Channel channel, CallOptions callOptions, T request) {
-    return new CallContext<T>(channel, callOptions, request);
+  public static <RequestT, ResponseT> CallContext<RequestT, ResponseT> of(
+      Channel channel, StreamObserver<ResponseT> responseObserver, RequestT request) {
```

Comment:
We're getting into a potential combinatorial explosion here. Might someone want to provide both CallOptions and StreamObserver?

Also, what should we do if someone provides us with a StreamObserver in a future call?

Is it worth using CallContext for both types of calls? (Maybe it is, but we should make sure we know why)